### PR TITLE
added fusing matmul-transpose-reshape pass

### PR DESCRIPTION
**Pull request by arlesniak**
This PR fuses the following sequence of ops for Ernie:
prev_op-(4dim)->mkldnn_matmul-(4dim)->transpose-(4dim)->reshape-(3dim)->next_op
------------------------------------------->  axis={0,2,1,3}---->shape={0,0,[-2]*[-1]}
into:
prev_op-(4dim)->mkldnn_matmul-(3dim)->next_op 

Also the fuse adds two new attributes to matmul_op:  fused_transpose_Out (axis from transpose) and fused_reshape_Out (shape from reshape) to be used by mkldnn MM kernel

### CODE REVIEW

**Comment by lidanqing-vv** [comment]
Hi @arlesniak  Please refer to [PADDLE_ENFORCE-Rewriting-Specification](https://github.com/PaddlePaddle/Paddle/wiki/PADDLE_ENFORCE-Rewriting-Specification) and change all the `PADDLE_ENFORCE` in your modified files, otherwise `PR_CI_CPU_Py2 (Compile)` will always fail. 

You can also refer to https://github.com/PaddlePaddle/Paddle/pull/23754#pullrequestreview-393223492

File: python/paddle/fluid/tests/unittests/ir/__init__.py
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/ir/__init__.py]: Are these two commented files needed?

File: python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py]: Do you consider using `check_output_with_option` here too, I saw you use it in later functions

File: python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py]: I think the op with `reshape_out` and `transpose_out` , should not mix up with op without these two attributes set. But in this PR, I did not see key changes, so it means, when reuse, you may use wrong cached op, because you did not add attribute value to the key.

Could you also add a cache mechanism check for the fused op. Simply speaking, you just need to run 2 iterations. You can refer Michal's matmul PR, I think he did this work.

File: python/paddle/fluid/tests/unittests/ir/__init__.py
Comment (arlesniak) [reply python/paddle/fluid/tests/unittests/ir/__init__.py]: Yes they allow to do the following import:  
> from ir.inference.inference_pass_test import InferencePassTest
in the test_matmul_op_output_fused.py

File: python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py
Comment (arlesniak) [reply python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py]: The `TestMatMulOpSpecial` class inherits from `OpTest` class which does not have the `check_output_with_option` method. The latter tests i.e. `TestMKLDNNMatmulFuseOp` inherit from `InferencePassTest` class there the method exists.

File: python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py
Comment (arlesniak) [reply python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py]: Concerning mixing op without attributes.
The test you mentioned is doing two runs of the little topology from `setUp`. First is raw inference pass, the second is with MKLDNN on and than they are compared. There is no need to put the attributes `reshape_out` and `transpose_out` to MKLDNN MM here because they are modified implicitly during the fuse pass.
Concerning caching:
Thank you for the comment I'll check this

Comment (arlesniak) [reply python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py]: @lidanqing-intel  OK I changed caching key and PADDLE_ENFORCE could you please see new version.

**Comment by arlesniak** [comment]
PR_CI_CPU_Py2 complains about only @skip_check_grad

> 10:22:32] :	 [Step 1/1] 0. It is an Op accuracy problem, please take care of it. You must have one RD (zhangting2020 (Recommend), luotao1 or phlrain) approval for the usage (either add or delete) of @skip_check_grad_ci. For more information, please refer to: https://github.com/PaddlePaddle/Paddle/wiki/Gradient-Check-Is-Required-for-Op-Test. The corresponding lines are as follows:
[10:22:32] :	 [Step 1/1] 
[10:22:32] :	 [Step 1/1] python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py
[10:22:32] :	 [Step 1/1] 
[10:22:32] :	 [Step 1/1] +	@skip_check_grad_ci(reason="Tests inference only optimization.")
[10:22:32] :	 [Step 1/1] 
[10:22:32] :	 [Step 1/1] 
[10:22:32] :	 [Step 1/1] There are 1 approved errors.

**Comment by arlesniak** [comment]
Please check new version

**Comment by lidanqing-vv** [comment]
LGTM

File: python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py
Comment (lidanqing-vv) [reply python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py]: Thanks for changing the key. LGTM.

File: paddle/fluid/framework/ir/mkldnn/matmul_transpose_reshape_fuse_pass_tester.cc
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/matmul_transpose_reshape_fuse_pass_tester.cc]: Please, verify also the attributes' values here.

File: paddle/fluid/framework/ir/mkldnn/matmul_transpose_reshape_fuse_pass_tester.cc
Comment (wojtuss) [inline paddle/fluid/framework/ir/mkldnn/matmul_transpose_reshape_fuse_pass_tester.cc]: Should the `reshape_Out` and `transpose_Out` attributes be set before applying the pass?

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (wojtuss) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:291]: A test with rank 3 `reshape_Out` is missing.

**Comment by sfraczek** [comment]
I would prefer not using transpose_Out name for attribute because there is transpose_X and transpose_Y already which have different meaning

File: paddle/fluid/framework/ir/mkldnn/matmul_transpose_reshape_fuse_pass_tester.cc
Comment (arlesniak) [reply paddle/fluid/framework/ir/mkldnn/matmul_transpose_reshape_fuse_pass_tester.cc]: OK

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (wojtuss) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:291]: A test with `int8` data type is also needed imo. This could be achieved by defining a class method
```
def init_data_type(self):
   self.data_type_ = 'float32'
```
then calling it in the setUp() method, and using self.data_type_ in the body of the `generate_data()` method instead of using the "float32" string. In the next class only override the init_data_type() method for INT8 test.

**Comment by wojtuss** [comment]
@luotao1, some builds failed due to exhausted space on a disc:
```
fatal error: error in backend: IO failure on output stream: No space left on device
```

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:127]: Batch size doesn't necessarily have to be defined in "X" tensor. This is the reason why it was inferred in such a non trivial way few lines above (`const memory::dim BS = x_bs || y_bs ? std::max(x_bs, y_bs) : 1;`)

In my opinion This should be rewritten as follows:
```C++
  if (BS > 1 && IsOutputFused(ctx)) {
      batch_size_ = BS;
      BS = 1;
  }
```
Alternatively
```
  if (BS > 1 && IsOutputFused(ctx)) {
      std::swap(batch_size_, BS);
  }
```
Variable `b` can therefore be omitted.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: Please consider renaming `correctStridesWhenOutputFused` to `CorrectStridesWhenOutputFused`

**Comment by GaoWei8** [comment]
> @luotao1, some builds failed due to exhausted space on a disc:
> 
> ```
> fatal error: error in backend: IO failure on output stream: No space left on device
> ```

The CIs have already rerun.   It will take more than 10 hours because of the long CI queue.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:207]: Since this for loop isn't parallelised, wouldn't it be more efficient to calculate each offset on the go?
This way we avoid a cache miss on fetching offsets_ table which could potentially cost us more cycles than just increasing `x_ptr`, `y_ptr ` and `out_ptr ` by the respective products of each memory dimensions (Please see the [documentation](https://oneapi-src.github.io/oneDNN/structdnnl_1_1memory.html#afdd20764d58c0b517d5a31276672aeb8)).
An exemplary implementation could look like this:
```C++
    unsigned bs = batch_size_;
    void* x_ptr = x_mem_.get_data_handle();
    void* y_ptr = y_mem_.get_data_handle();
    void* out_ptr = out_mem_.get_data_handle();
    char* x_size = x_mem_.dims[1] * x_mem_.dims[2];
    char* y_size = y_mem_.dims[1] * y_mem_.dims[2];
    char* out_size = out_mem_.dims[1] * out_mem_.dims[2];
    for (unsigned i = 0; i < bs; i++) {
      x_mem_.set_data_handle(x_ptr);
      y_mem_.set_data_handle(y_ptr);
      out_mem_.set_data_handle(out_ptr);
      matmul_prim_.execute(stream, {
                                       {MKLDNN_ARG_SRC, x_mem_},
                                       {MKLDNN_ARG_WEIGHTS, y_mem_},
                                       {MKLDNN_ARG_DST, out_mem_},
                                   });
      x_ptr = static_cast<char*>(x_ptr) + x_size;
      y_ptr = static_cast<char*>(y_ptr) + y_size;
      out_ptr = static_cast<char*>(out_ptr) + out_size;
    }
```
This would also reduce cache trashing as we reduce the size of the class by removing the offsets_ field.

File: paddle/fluid/framework/ddim.h
Comment (Sand3r-) [inline paddle/fluid/framework/ddim.h]: Since the header file only attempts to implement template and inline methods, perhaps it would be more suited to move the implementations of `reshape` and `transpose` to .cpp file

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: Please, correct me if I'm wrong, but I have an impression that the axis stored in `fused_transpose_Out` is currently completely disregarded in the code of dnnl's matmul, while as far I understand it, it influences how the strides are laid out. Perhaps it'd be advised to ensure with paddle enforce mechanism what transpose axis is currently supported.

File: paddle/fluid/framework/ddim.h
Comment (Sand3r-) [inline paddle/fluid/framework/ddim.h]: `The index of 0 in `shape` must be less than the dimensions.`

Could you please clarify the message? @sfraczek Could you help here?
My best guess is that "Index %d of shape under which the value of 0 is stored, must be lower than the **number** of old dimensions." was meant.

**Review by Sand3r- — COMMENTED** [review]
@arlesniak thank you so much for all the hard work you've spent on the PR. Good job in general, I really appreciate the UTs you've prepared, as well as keeping the solution clean in general. Please refer to my comments.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: I see it used for shape_out but its a different file

https://github.com/PaddlePaddle/Paddle/pull/23866/files/d93bb88437fbbb690d11338e80f36707cec3750a#diff-126803a76a1b0cbf6198f30e23058f8dR419-R436

Comment (arlesniak) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: It's used in the Op not the kernel

Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: That's right, thank you for clarification.

File: paddle/fluid/framework/ddim.h
Comment (sfraczek) [reply paddle/fluid/framework/ddim.h]: Sounds good to me. We can take your suggestion.

Comment (sfraczek) [reply paddle/fluid/framework/ddim.h]: @Sand3r-  Should we keep the part "But received shape[...."?

Comment (Sand3r-) [reply paddle/fluid/framework/ddim.h]: Oh yes, totally :)

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:207]: it doesn't compile. 
` error: invalid use of ‘dnnl::memory::dims’`

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:127]: Why should we change batch size to 1 then?

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:207]: Anyway, matmul didn't work when we multiplied batch size with 3rd dim so I guess batch 1 will also not work

Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:207]: Checking this modification would add +1 / +2 weeks to development and testing

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (arlesniak) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: done

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (arlesniak) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py:291]: done

File: paddle/fluid/framework/ddim.h
Comment (arlesniak) [reply paddle/fluid/framework/ddim.h]: done

File: paddle/fluid/framework/ddim.h
Comment (arlesniak) [reply paddle/fluid/framework/ddim.h]: done

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:127]: To set 1 as the first entity in the shape of memory descriptors. This is what's done here anyway isn't it?:
```
b = BS / batch_size_;
```
After all BS is the same thing as batch_size_.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:207]: > it doesn't compile.
> ` error: invalid use of ‘dnnl::memory::dims’`

It might be, I haven't compiled this but attempted to give a brief idea of what I mean, hence "An exemplary implementation could look like this:" disclaimer. 

> Anyway, matmul didn't work when we multiplied batch size with 3rd dim so I guess batch 1 will also not work

Hmm I don't quite understand how does this refer to the proposal? Nothing regarding batch size is changed here.

**Comment by lidanqing-vv** [comment]
@luotao1  Please approve this PR_CI_CPU_Py2, it fail because of check_grad_ci. We want to cherry-pick it to 2.0

File: paddle/fluid/operators/matmul_op.cc
Comment (wojtuss) [inline paddle/fluid/operators/matmul_op.cc]: IMO if we check with enforce for supported transpose axis and reshape out size here, we should check these conditions in the fuse pass before executing it (now we apply the fuse unconditionally). With the current version, it should work fine with the Ernie model, but can lead to failure with some other model.

**Comment by luotao1** [comment]
Please update your PR description.

File: paddle/fluid/operators/matmul_op.cc
Comment (luotao1) [inline paddle/fluid/operators/matmul_op.cc:498]: Is this necessary to add two attributes in matmul_op? And the comments are not enough, maybe you can improve the comments in the next PR?

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:127]: let's discuss this change after its merged. I can do proper testing then to verify if that change would still work. At least regarding my PR. Artur's PR is up to him but he probably feels the same on this.

File: paddle/fluid/operators/matmul_op.cc
Comment (sfraczek) [reply paddle/fluid/operators/matmul_op.cc:498]: It would be nice to see a pair of vectors for example because we require both attributes present anyway. Hower, we didn't want to extend the Attribute definition or hack it. Should we?
https://github.com/PaddlePaddle/Paddle/blob/a31d7328b70fde9c860dfad918222d7b2d842d71/paddle/fluid/framework/type_defs.h#L41-L45

Comment (arlesniak) [reply paddle/fluid/operators/matmul_op.cc:498]: I can improve comments in the next PR

**Review by luotao1 — DISMISSED** [review]
LGTM

**Comment by luotao1** [comment]
> Please approve this PR_CI_CPU_Py2

@lidanqing-intel Please have any other LGTM at first.

**Comment by arlesniak** [comment]
@luotao1 I have improved comments and review fixes ready. Should I push them now or they can wait till next PR ?

File: paddle/fluid/operators/matmul_op.cc
Comment (arlesniak) [reply paddle/fluid/operators/matmul_op.cc]: Done

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:207]: @sfraczek and I have figured out that it's actually not possible to access the dimensions directly from memory primitive. 
Instead we'd need to fetch the memory descriptor and then from it the actual dimensions which are represented by a vector returned by value. 
Such copying of vectors and multiple levels of indirection would actually result in slower execution.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:127]: After a discussion with @sfraczek we've came to an understanding on how should this work. 
We've figured out that the batch_size_ should actually be assigned that way:
```C++
auto& x_dims = ctx.Input<Tensor>("X")->dims();
auto& y_dims = ctx.Input<Tensor>("Y")->dims();
batch_size_ = x_bs > y_bs ? x_dims[0] : y_dims [0];
```

The reset can stay the same.

Comment (arlesniak) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc:127]: @Sand3r- Doesn't the prior ENFORCE_EQ (line 112) rules the conditions earlier, so we don;t have to be worried about eqaulity of the x, y batches ?
 PADDLE_ENFORCE_EQ(x_bs > 0 && y_bs > 0 && x_bs != y_bs, false,
                      platform::errors::InvalidArgument(
                          "If batch sizes of X and Y are positive,"
                          "they have to be equal."));

**Review by wojtuss — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]

**Comment by wojtuss** [comment]
@luotao1 , is there still time to cherry-pick this PR to 2.0 release?

**Comment by luotao1** [comment]
Yes, you can cherry-pick to 2.0 release

**Pull request merged**
